lib/core/src/Cardano/Wallet/Api/Types/BlockHeader.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE RecordWildCards #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Wallet.Api.Types.BlockHeader where
+
+import Prelude hiding
+    ( id )
+
+import Cardano.Wallet.Api.Lib
+    ( fromTextJSON, toTextJSON )
+import Cardano.Wallet.Primitive.Types
+    ( BlockHeader (..), SlotNo (..) )
+import Cardano.Wallet.Primitive.Types.Hash
+    ( Hash (..) )
+import Data.Aeson
+    ( FromJSON (parseJSON), ToJSON (toJSON) )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Word
+    ( Word32, Word64 )
+import GHC.Generics
+    ( Generic )
+
+instance FromJSON (Hash "BlockHeader") where
+    parseJSON = fromTextJSON "Block Header Hash"
+
+instance ToJSON (Hash "BlockHeader") where
+    toJSON = toTextJSON
+
+data ApiBlockHeader = ApiBlockHeader
+  { id :: Hash "BlockHeader",
+    slot :: Quantity "slot" Word64,
+    height :: Quantity "block" Word32
+  }
+  deriving (Eq, Show, Generic, ToJSON, FromJSON )
+
+mkApiBlockHeader :: BlockHeader -> ApiBlockHeader
+mkApiBlockHeader BlockHeader {..} = ApiBlockHeader {..}
+  where
+    id = headerHash
+    height = blockHeight
+    slot = Quantity $ unSlotNo slotNo
```

Comment:
```suggestion
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE RecordWildCards #-}

module Cardano.Wallet.Api.Types.BlockHeader where

import Prelude

import Cardano.Wallet.Api.Lib
    ( fromTextJSON, toTextJSON )
import Cardano.Wallet.Primitive.Types
    ( BlockHeader (..) )
import Data.Aeson
    ( FromJSON (parseJSON), ToJSON (toJSON), (.:), (.:?), (.=) )

import qualified Data.Aeson as Aeson

newtype ApiBlockHeader = ApiBlockHeader BlockHeader
    deriving newtype (Eq)
    deriving stock (Show)

instance ToJSON ApiBlockHeader where
    toJSON (ApiBlockHeader BlockHeader{..}) =
        Aeson.object
            [ "slotNo" .= slotNo
            , "blockHeight" .= blockHeight
            , "headerHash" .= toTextJSON headerHash
            , "parentHeaderHash" .= fmap toTextJSON parentHeaderHash
            ]

instance FromJSON ApiBlockHeader where
    parseJSON = Aeson.withObject "ApiBlockHeader" $ \o -> do
        slotNo <- o .: "slotNo"
        blockHeight <- o .: "blockHeight"
        headerHash <- o .: "headerHash"
            >>= fromTextJSON "headerHash"
        parentHeaderHash <- o .:? "parentHeaderHash"
            >>= traverse (fromTextJSON "parentHeaderHash")
        pure . ApiBlockHeader $ BlockHeader {..}
```

Reply:
Why I like green solution more:

1. It uses 0-cost newtype.
2. It doesn't involve orphaned instances.
3. It allows to decouple object field names in JSON from the object field names in Haskell. This way I have a freedom to refactor Haskell code without breaking API contract. 
4. The parent header hash is also in the JSON.